Use ViewChangesAsync consistently for changes

gerhardol opened the pull request:
Part of #7338 
Related to #7825 
Fixes #7866 

## Proposed changes

There are many codepaths that "presents changes" to files (binary, images) and submodules. The presentation differs if the file is new, tracked etc.
The logic for that is split between the separate forms and GitUIExtensions.cs, ViewChangesAsync(). Some of that has been cleaned up in previous PRs.
(FormCommit.cs also needs to have some logic in FileViewer.cs ViewCurrentChanges(), that requires #7825 first.)
The presentation therefore differs from form to form.

This change unifies the handling to be consistent, all using ViewChangesAsync()

This PR shares (conflicts) a change in #7914, rev-parse is required to get guid, see the explanation there.

This only changes how FormStash presents changes, the only actual change of this PR is that images that are changed should now be displayed.
(and FormCommit should not see any changes when this is applied in or after #78825, depending what goes in first).

## Test methodology
Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv commented:
@gerhardol, could you create a list of your PRs with their dependencies, please, in order to know which to review first.

gerhardol commented:
> could you create a list of your PRs with their dependencies

Some kond of review order, with simple and obvious first

#7926 refactoring and independent functionality but conflict with #7914 . #7825 needed to complete this (I expect this to go in first and #7825 be updated) 
#7914 refactoring and independent functionality but conflict with #7926

#7900 No major dependencies, to be updated soon

#7922 No dependencies

#7927 independent (and simple)

#7923 Move out the "algorithm" part of #7899, to at least start the discussion about the change

#7825 A few relations to files above, other than #7825 only rebase

#7899 Refactoring for GitItemStatusWithParents, enables unique diff of BASE->first. After #7923 right now. A major blocker for me.

#7915 WIP for discussion of how to handle parse errors. Need to be rebased on #7926

TBD
fix FileViewer API: Use GitItemStatusWithParents from #7899 and #7914 to cleanup the API in #7825

mstv commented on GitUI/GitUIExtensions.cs:
```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
+
+                    return string.IsNullOrWhiteSpace(diffOfConflict)
+                        ? diffOfConflict
+                        : Strings.UninterestingDiffOmitted;
```
```suggestion
                    return string.IsNullOrWhiteSpace(diffOfConflict)
                        ? Strings.UninterestingDiffOmitted
                        : diffOfConflict;
```

gerhardol replied:
ok

mstv commented on GitUI/GitUIExtensions.cs:
```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
+                    var status = ThreadHelper.JoinableTaskFactory.Run(file.GetSubmoduleStatusAsync);
+                    return status != null
+                        ? LocalizationHelpers.ProcessSubmoduleStatus(fileViewer.Module, status)
+                        : $"Submodule \"{file.Name}\" has unresolved conflicts";
```
Consider localization. Perhaps rather support `null` in `ProcessSubmoduleStatus`.

gerhardol replied:
I cannot see how the submodule status would not be evaluated except in case of errors, so I changed the message and suggest no translation.

mstv commented on GitUI/GitUIExtensions.cs:
```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
-                    return LocalizationHelpers.GetSubmoduleText(diffViewer.Module, file.Name.TrimEnd('/'), "");
+                    return fileViewer.ViewTextAsync(file?.Name, defaultText, openWithDiffTool);
                 }
-            }
```
Can the whole block `if (!file.IsTracked)` be omitted safely?

gerhardol replied:
Yes, !IsTracked -> IsNew is always set and we never try to get patches
This is handled in FileViewr.cs ViewFileAsync():397 (also for other code paths), see also #7869

gerhardol commented on GitUI/CommandsDialogs/FormStash.cs:
```diff
@@ -187,52 +188,22 @@ private void StashedSelectedIndexChanged(object sender, EventArgs e)
-                            });
-                    }
+                    first = ObjectId.IndexId;
+                    selected = ObjectId.WorkTreeId;
```
> Though I guess stash stores both unstaged and staged files, doesn't it?

Yes
This causes the worktree files to be displayed in the viewer rather than the index, so incorrect...
Will fix the versions
(This is better handled with GitItemStatusWithParent after #7899)

gerhardol replied:
worktree and index should be presented in two groups too (this definitely requires #7899)

gerhardol replied:
Improved the change to only rev-parse when creating the FileStatusList and separated to own commit - could be a separate PR (but still dependent on this PR).

mstv commented on GitUI/UserControls/FileStatusList.cs:
```diff
@@ -726,12 +726,16 @@ ObjectId GetRevisionOrHead(ObjectId rev, Lazy<ObjectId> head)
+            GitItemStatusesWithDescription = new[] { (parentRev, _diffWithParent.Text + GetDescriptionForRevision(parentRev?.ObjectId), items) };
+        }
+
+        public void SetDiffs()
```
Should it rather be renamed to `ClearDiffs` or `ResetDiffs`?

gerhardol replied:
ClearFileStatus()
I did not want to change so much in this PR - will change this and open a new PR for Stash

gerhardol replied:
It should be easier to get smaller PRs merged, but with @RussKie you never know. The smaller the PRs are, the more changes requested per line changed. There is a review queue...
But many of my open PRs are related to each other, I am blocked and will probably spend time on other projects now.

gerhardol replied:
> But the counterpart is named SetDiffs

The method could maybe be called Clear() instead
SetDiffs() is not a good name, but I do not have any better suggestion

gerhardol replied:
As mentioned in #7938, changing to ClearDiffs()

mstv commented on IntegrationTests/UI.IntegrationTests/CommandsDialogs/FileStatusListTests.cs:
```diff
@@ -48,7 +48,7 @@ public void ItemSelections()
             var itemAt1 = item2;
             var itemAt2 = item0;
-            _fileStatusList.SetDiffs(items: items);
+            _fileStatusList.SetDiffs(null, null, items: items);
```
Yes, Re# injects these -- but it slows down the IDE too much for me.
```suggestion
            _fileStatusList.SetDiffs(selectedRev: null, parentRev: null, items);
```

gerhardol replied:
ok, for all

mstv commented on GitUI/CommandsDialogs/RepoHosting/ViewPullRequestsForm.cs:
```diff
@@ -367,7 +367,9 @@ private void SplitAndLoadDiff(string diffData)
-            _fileStatusList.SetDiffs(items: giss);
+            // Commits in PR may not exist in the repo so GitRevision in FileStatusList is not used,
+            // patches directly from the cache
+            _fileStatusList.SetDiffs(null, null, items: giss);
```
```suggestion
            _fileStatusList.SetDiffs(selectedRev: null, parentRev: null, items: giss);
```

gerhardol commented:
Removed the FormStash improvements, will create a new PR continued from this, also likely separating worktree/index in groups.

gerhardol commented:
@RussKie I would like to merge this and you have approved the follow up #7938. OK?

RussKie commented:
I'll review this today

RussKie commented on GitCommands/FileHelper.cs:
```diff
@@ -57,8 +57,9 @@ public static class FileHelper
         {
-            return IsBinaryAccordingToGitAttributes(module, fileName)
-                ?? HasMatchingExtension(BinaryExtensions, fileName);
+            return fileName.IsNotNullOrWhitespace()
+                   && (IsBinaryAccordingToGitAttributes(module, fileName)
+                       ?? HasMatchingExtension(BinaryExtensions, fileName));
```
Do we have tests for this?

gerhardol replied:
No

RussKie commented on GitUI/GitUIExtensions.cs:
```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
+        /// <param name="defaultText">default text if no diff is possible</param>
+        /// <param name="openWithDiffTool">The difftool command to open with</param>
+        /// <returns>Task to view</returns>
+        public static Task ViewChangesAsync(this FileViewer fileViewer,
```
If we make it `this IFileViewer fileViewer`, we should be able to test for of the codepaths

gerhardol replied:
There is no interface for the methods FileViewer.cs provides right now (like ViewTextAsync(), ViewGitItemRevisionAsync(), Clear()).
The IFileViewer interface handles FileViewerInternal.cs only right now (but is not used externally.)
This would also require that Module is supplied separetly (or difftool handling is removed as discussed).
So not an obvious change.

RussKie replied:
Ah, I remembered we had `IFileViewer`, it was on a different type...

RussKie commented on GitUI/GitUIExtensions.cs:
```diff
@@ -34,82 +34,90 @@ public static class GitUIExtensions
-                // View blob guid from revision, or file for worktree
-                return diffViewer.ViewGitItemRevisionAsync(file, selectedRevision.ObjectId, openWithDifftool);
-            }
+                if (file.IsSubmodule && file.GetSubmoduleStatusAsync() != null)
```
`file.GetSubmoduleStatusAsync() != null` this looks sus. An async method should be awaited, not checked for `null`s...

gerhardol replied:
This pattern is used in many situations (this PR reduces the usage, some more with #7825).
Submodule status is not always evaluated, this checks if the async Task has been started, if so await the result on the next line. (This is the normal code path I believe.)
If not started async, the submodule status is evaluated below.

gerhardol commented:
> My biggest worry with sweeping changes like this is testing. There are lots of changes not accompanied by tests.

https://github.com/gitextensions/gitextensions/pull/7926#pullrequestreview-385057653

Currently, there are six different ways to present changes, with this PR this is down to two (with #7825 change to be added after this PR there will be one).
The different implementations is part of the form code and requires a refactoring anyway to be tested.
Also, I cannot see any sane tests that could test test the migration.

gerhardol commented:
Rebased and added an update in relation to https://github.com/gitextensions/gitextensions/pull/7938#discussion_r403334066

Moving out logic to handle ViewPatch as text from FileViewer, to keep the logic in the ViewChangesAsync extension method

FileViewer still has logic to detect binary files, to determine how to display "git items" (image, submodules or text) and to display .patch files as patches (diffs).

gerhardol commented:
@RussKie further comments?

RussKie commented:
I'll have a look today.

gerhardol commented:
@msftbot merge in 12 hours
